pallets/score-staking/src/lib.rs

```diff
@@ -0,0 +1,419 @@
+		RoundConfigSet { new_config: RoundSetting },
+		ScoreUpdated { who: Identity, new_score: Score },
+		ScoreRemoved { who: Identity },
+		ScoreCleared {},
+		RewardCalculated { total: BalanceOf<T>, distributed: BalanceOf<T> },
+		RewardClaimed { who: T::AccountId, amount: BalanceOf<T> },
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn score_feeder)]
+	pub type ScoreFeeder<T: Config> = StorageValue<_, T::AccountId, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn round)]
+	pub type Round<T: Config> = StorageValue<_, RoundInfo<BlockNumberFor<T>>, ValueQuery>;
+
+	#[pallet::type_value]
+	pub fn DefaultRoundSetting<T: Config>() -> RoundSetting {
+		RoundSetting {
+			interval: DEFAULT_ROUND_INTERVAL,
+			score_coefficient: DEFAULT_SCORE_COEFFICIENT,
+			stake_coefficient: DEFAULT_STAKE_COEFFICIENT,
+		}
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn round_config)]
+	pub type RoundConfig<T: Config> =
+		StorageValue<_, RoundSetting, ValueQuery, DefaultRoundSetting<T>>;
+
+	// use `Twox64Concat` and `T::AccountId` for faster and shorter storage
+	// we might have tens of thousands of entries
+	#[pallet::storage]
+	#[pallet::getter(fn scores)]
+	pub type Scores<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, ScorePayment<BalanceOf<T>>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn state)]
+	pub type State<T: Config> = StorageValue<_, PoolState, ValueQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub state: PoolState,
+		pub marker: PhantomData<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self { state: PoolState::Stopped, marker: Default::default() }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			State::<T>::put(self.state);
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(now: BlockNumberFor<T>) -> Weight {
+			let mut weight = T::DbWeight::get().reads_writes(1, 0); // Self::state()
+
+			if Self::state() == PoolState::Stopped {
+				return weight
+			}
+
+			let mut r = Round::<T>::get();
+			weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 0));
+
+			if !is_modulo(now - r.start_block, Self::round_config().interval.into()) {
+				// nothing to do there
+				return weight
+			}
+
+			// We are about to start a new round
+			// 1. update round info
+			r.index = r.index.saturating_add(1);
+			r.start_block = now;
+			Round::<T>::put(r);
+			weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+			// 2. calculate payout
+			let round_reward: BalanceOf<T> = (T::YearlyInflation::get() * T::YearlyIssuance::get() /
+				YEARS.into()) * Self::round_config().interval.into();
+
+			let total_stake = ParaStaking::Pallet::<T>::total();
+			let total_score =
+				Scores::<T>::iter_values().fold(0u32, |a, b| a.checked_add(b.score).unwrap_or(a));
```

Comment:
Maybe instead of iterating, have an extra storage value to keep trace of total score?

Reply:
I can do it

Reply:
Improved